great_expectations/zep/metadatasource.py

```diff
@@ -3,155 +3,59 @@
-from great_expectations.zep.interfaces import Datasource
+from great_expectations.zep.context import _SourceFactories
 
-SourceFactoryFn = Callable[..., Datasource]
+if TYPE_CHECKING:
+    from great_expectations.zep.interfaces import Datasource
 
+SourceFactoryFn = Callable[..., "Datasource"]
 
-def _remove_suffix(s: str, suffix: str) -> str:
-    # NOTE: str.remove_suffix() added in python 3.9
-    if s.endswith(suffix):
-        s = s[: -len(suffix)]
-    return s
+LOGGER = logging.getLogger(__name__)
 
 
-class _SourceFactories:
-
-    __source_factories: Dict[str, SourceFactoryFn] = {}
+class MetaDatasource(type):
+    def __new__(
+        meta_cls: Type[MetaDatasource], cls_name: str, bases: tuple[type], cls_dict
+    ) -> MetaDatasource:
+        """
+        MetaDatasource hook that runs when a new `Datasource` is defined.
+        This methods binds a factory method for the defined `Datasource` to `_SourceFactories` class which becomes
+        available as part of the `DataContext`.
 
-    @classmethod
-    def register_factory(
-        cls,
-        ds_type: type,
-        fn: SourceFactoryFn,
-    ) -> None:
+        Also binds asset adding methods according to the declared `asset_types`.
         """
-        Add/Register a datasource factory function.
+        LOGGER.info(f"1a. {meta_cls.__name__}.__new__() for `{cls_name}`")
 
-        Derives a SIMPLIFIED_NAME from the provided `Datasource` type.
-        Attaches a method called `add_<SIMPLIFIED_NAME>()`.
+        # TODO: extract asset type details to build factory method signature etc. (pull args from __init__)
 
-        Example
-        -------
-        `class PandasDatasource` -> `add_pandas()`
-        """
-        simplified_name = _remove_suffix(
-            camel_to_snake(ds_type.__name__), "_datasource"
-        )
-        method_name = f"add_{simplified_name}"
-        print(
-            f"2. Registering {ds_type.__name__} as {simplified_name} with {method_name}() factory"
-        )
-
-        pre_existing = cls.__source_factories.get(method_name)
-        if not pre_existing:
-            cls.__source_factories[method_name] = fn  # type: ignore[assignment]
-        else:
-            raise ValueError(f"{simplified_name} factory already exists")
-
-    @property
-    def factories(self) -> List[str]:
-        return list(self.__source_factories.keys())
-
-    def __getattr__(self, name):
-        try:
-            return self.__source_factories[name]
-        except KeyError:
-            raise AttributeError(name)
-
-    def __dir__(self) -> List[str]:
-        return ["register_factory", "factories", *self.factories]
-
-
-class MetaDatasouce(type):
-    def __new__(meta_cls, cls_name, bases, cls_dict) -> MetaDatasouce:
-        print(f"1. {meta_cls.__name__}.__new__() for `{cls_name}`")
+        asset_types: List[type] = cls_dict.get("asset_types")
+        LOGGER.info(f"1b. Extracting Asset details - {asset_types}")
+        if asset_types:
```

Comment:
I think we can remove this check now that we know `asset_types` is a `List[Type[DataAsset]]`?

Reply:
Yeah, we can rely on the type-checker for this.